bench/macro/lsm-tree-bench-unions.hs

```diff
@@ -0,0 +1,629 @@
+          unions = Set.unions tKeySets
+          uniques = Fold.foldl' (\x -> Set.union (gatherUniques x)) mempty tKeySets
+      pure $ unions `Set.difference` uniques
+
+    printf "True duplicates: %d\n" $ Set.size duplicates
+
+-- |
+-- From StackOverflow: https://stackoverflow.com/a/30938328
+renderRational :: Int -> Rational -> String
+renderRational len rat = sign <> shows prefix ("." ++ suffix)
+    where
+      sign
+        | num < 0 = "-"
+        | otherwise = ""
+
+      (prefix, next) = abs num `quotRem` den
+
+      suffix = case next of
+        0 -> "0"
+        n -> take len $ go n
+
+      num = numerator rat
+      den = denominator rat
+      go 0 = ""
+      go x = let (d', next') = (10 * x) `quotRem` den
+             in shows d' (go next')
+
+-------------------------------------------------------------------------------
+-- PRNG initialisation
+-------------------------------------------------------------------------------
+
+initializeRNG :: GlobalOpts -> IO ()
+initializeRNG gopts =
+    -- Create an RNG for randomized insertion of entries
+    case seed gopts of
+        -- When no seed was explicitly supplied, used system entropy.
+        -- This is the default semantics of the @StdGen@ data-type,
+        -- hence no action in required.
+        Nothing           -> pure ()
+        -- When the seed was explicitly specified by the user,
+        -- overide the system RNG with the explict seed to ensure determinism.
+        Just explicitSeed -> Random.setStdGen $ Random.mkStdGen explicitSeed
+
+deriveMultipleRNGs :: Int -> IO (NonEmpty Random.StdGen)
+deriveMultipleRNGs amount | amount <= 0 = fail "Must specify a POSITIVE number of RNGs to request"
+deriveMultipleRNGs amount =
+   let splitContinuation 0 gens done = (done, gens)
+       splitContinuation n gens prev =
+              let (curr, next) = Random.splitGen prev
+              in  splitContinuation (n - 1) (curr <| gens) next
+   in  do  rng0 <- Random.getStdGen
+           let (rng1, rng2) = Random.splitGen rng0
+               (rngN, splitRNGs) = splitContinuation amount (rng1 :| []) rng2
+           Random.setStdGen rngN
+           pure splitRNGs
+
+
+-------------------------------------------------------------------------------
+-- Sampling generation & measurements
+-------------------------------------------------------------------------------
+
+-- | The tuple of measured timings over a sample series @(mean, min, max, variance)@.
+type TimingResult = (Integer, Integer, Integer, Integer)
+
+-- | A single batch to be sampled as part of a sampling series for a measurement.
+{-# INLINE generateBatchSampling #-}
+generateBatchSampling  ::
+       Int       -- ^ initial size of the collection
+    -> Int       -- ^ batch size
+    -> IO (V.Vector K)
+generateBatchSampling initialSize batchSize = toOperations <$> V.replicateM batchSize randomKey
+  where
+    randomKey :: IO Word64
+    randomKey = Random.randomRIO (0, 2 * fromIntegral initialSize - 1)
```

Comment:
I'd advise to pass around RNGs purely. That would make `generateBatchSampling` look something like this:
```suggestion
-- | A single batch to be sampled as part of a sampling series for a measurement.
{-# INLINE generateBatchSampling #-}
generateBatchSampling  ::
       Int       -- ^ initial size of the collection
    -> Int       -- ^ batch size
    -> Random.StdGen
    -> (V.Vector K, Random.StdGen)
generateBatchSampling initialSize batchSize !g0 =
    runState (V.replicateM batchSize (state randomKey)) g0
  where
    randomKey :: Random.StdGen -> (K, Random.StdGen)
    randomKey !g =
      let (!x, !g') = Random.randomR (0, 2 * fromIntegral initialSize - 1) g
          !k = makeKey x
      in  (k, g')
```